### ISSUE TITLE
Only pretranslate localizable text in complex Fluent messages

### ISSUE THREAD

**Issue by mathjazz** [mozilla/pontoon#2733]
Pretranslation rarely translates complex Fluent messages (messages with selectors, messages with multiple attributes, messages with a value and an attribute...) correctly, because it breaks Fluent syntax.

We should fix that by only translating localizable parts of messages. To quote Eemeli from the [last week](https://github.com/mozilla/pontoon/pull/2716#discussion_r1082192660):

> We really ought to extract each flat pattern, and have the ability to re-integrate it in its context. Consider for instance the following real-world message:
> 
> ```ftl
> detail-show-preferences =
>     .label =
>         { PLATFORM() ->
>             [windows] Options
>            *[other] Preferences
>         }
>     .tooltiptext =
>         { PLATFORM() ->
>             [windows] Change this add-on’s options
>            *[other] Change this add-on’s preferences
>         }
> ```
> 
> Here, we should be able to find and separate the four patterns in this message, translate each separately, and then feed them back into the message structure at the appropriate slots.

**Comment by mathjazz** [comment]
Here's a high-level overview of the algorithm we intend to use to turn the original string into a pretranslated string:

1. For message value and each attribute, uplift any selectors to the highest possible level and duplicate shared parts in the variants accordingly.
2. Flatten each value, attribute and variant that consists of multiple elements.
3. Replace each flat value, attribute and variant with its pretranslation.

**Comment by eemeli** [comment]
For context, this is how we did it in the front-end when it was using the Fluent AST: https://github.com/mozilla/pontoon/blob/26c5f709ce9ddec66ec214c4ec8db722196ea24c/translate/src/utils/message/flattenPatternElements.ts

That also mostly flattened messages with more than one selector, but those should probably just be left out at least of the first pass at applying a similar transform for pretranslation.

**Comment by mathjazz** [comment]
Should I not use https://github.com/mozilla/pontoon/blob/85047a2/translate/src/utils/message/parseFlatFluent.ts as reference?

**Comment by eemeli** [comment]
Yeah, the earlier version that I linked to also does the uplifting; that part of the comment is now out of date. The uplifting is now effectively handled by the MF2 data model transform.

**Comment by mathjazz** [comment]
Oh, right. Thanks for clarifying!
